### ISSUE TITLE
Snapshots system for managing change in datasets

### ISSUE THREAD

**Issue by jywarren** [publiclab/spectral-workbench#192]
Complete!!! See documentation: #243

**Less urgent:**

* [ ] Eventually, we could deprecate the `spectrum.data` column entirely in favor of data snapshots in their own table/model. Or `spectrum.data` would become the "archived" never-touch-it original version.

**Wontfix:**

* [ ] ensure tag order with a submitted created_at timestamp from the client side? Needed?
<strike>* [ ] should we use a migration to pre-generate snapshots for any existing powertags? **YES**, we have to -- so that when people reference a spectrum, they can simply call the most recent snapshot and clone it. Otherwise there's no easy way to generate usable snapshots on server side based on a tag being created -- we don't want to have to enclose the data in the tag creation request. 
  * but CRAP! we can't do that in a migration, then? Must think on this. **OK, won't do it.**</strike>

**Complete:**

* [x] figure out what operations cause snapshotting:
  * [x] manual calibration (what's the tag?)
  * [x] calibrate
  * [x] subtract
  * [x] transform
  * [x] range
  * [x] crossSection
  * [x] smooth:
* [x] use snapshot instead of overwriting original spectrum in:
  * [x] calibration (server side)
  * [x] clone calibration (moved to client side)
    * [x] check that it actually copies the calibration
    * [x] check that it works after a range tag
  * [x] set sample row (server side) / extract
  * [x] range (just metadata, not actual data deletion)
  * [x] refactor `$.ajax` based spectrum fetching to an API call, which is snapshot-aware, using `snapshot_id` and latest/original defaults (or error if # refers to non-existent snapshot)
* [x] consider if qualitative tags like `calibrationQuality` or `error` should be deleted/updated to reflect most recent snapshot/calibration? We'll say YES for now. Tags describe current state.
* [x] de-conflict use of `#` in blend powertag, check site for usage of this tag. Substitute "$" 
  * [x] update tests (there are none for `blend`)
  * [x] update validations
* [x] following are almost done, but tag.parse() runs after tag.upload(), so we don't have the data ready at upload time
  * [x] ok, this is done, but now we should run tests and check that synchronous tag parsing hasn't mucked anything up, 
  * [x] also our Jasmine tests should be easier and rewriteable
* [x] adjust client side tag name to include snapshot id on response
  * [x] double-check client receipt of new #name! or just do it if you point at a spectrum and need to get 
* [x] beware that admins can add tags that aren't by the same user as the spectrum... could that cause trouble?
  * [x] admins making snapshots could make them on behalf of the author, with their user_id?
* [x] what about manual calibration? It should generate a snapshot too, 
* [x] and not overwrite the original...
* [x] any reference to a spectrum by a powertag auto-generates a Snapshot of the referred-to spectrum, owned by the tag creator
* [x] ensure that any calibration cloning causes tag generation
* [x] will #s disrupt client-side tag parsing? No.
* [x] do snapshots have a tag_id, or do tags have a snapshot_id? if tags have snapshot_id, we can display an icon easily, and maybe skip the `#1234` notation
* [x] tags and snapshots should be 1:1, which will create a lot of snapshots if a spectrum is referred to a lot, but that's OK
* [x] snapshots are ordered in time?
* [x] what if the tags used to generate the snapshot are deleted, and the snapshot is "hanging" outside the timeline? (this can't happen now, because of below)
* [x] snapshots are deleted if their powertags are deleted
  * [x] only if there is only one tag referring to it
  * [x] create `snapshot.tags` finder, and allow multiple tags to refer to a single snapshot
* [x] what about copy calibration? Doesn't need it, but let's do it anyways.
* [x] powertags refer to specific snapshots with syntax `subtract:1234#6343` -- wordy, but worth it for the unique id. Better than versioning because we can insert between two snapshots.
* [x] powertagging interface must default to latest snapshot, or offer ability to choose
* [x] can tags by multiple people refer to the same snapshot? Or only your own snapshots? (only 1:1 tags/snapshots)
* [x] powertags default to current/latest moment
* [x] we are mixing up #notation! it's implemented pointing at the snapshot **caused by** the tag, but it should be the snapshot **relied upon** by the tag
  * [x] does this mean we do need a snapshot.tag_id after all? Yes!
  * [x] also, if a spectrum has no snapshots, we can just point at the original
* [x] snapshot data is calculated on the client side upon tag creation, and enclosed with the submitted tags.
  * [x] if multiple tags are submitted, like "smooth:8,subtract:14", we need to enclose data for each. 
  * [x] break up the tags into separate requests, and snapshot each -- better consistent than fast. 
* [x] WHAT STOPS a user from manually adding a snapshot_id via the tagname, then deleting the snapshot once the tag is deleted?
  * [x] should we just not delete snapshots? or validate tagnames and disallow manual snapshot association, perhaps through spectrum.add_snapshot?
  * No, this is not a problem -- we AREN'T (or aren't supposed to be) using the tag.name to indicate the tag's generated snapshot. Just to indicate a snapshot a tag POINTS at. 
* [x] spectrum_id, user_id, description, data, timestamps
* [x] can be accessed via `spectrum.latest` and `spectrum.snapshots`
* [x] Unit tests! 
* [x] Must we upload spectrum data from client side upon snapshot generation? Yes.
* [x] Reject snapshot tags which don't enclose data? No.
* [x] no deleting snapshots if **anything** depends on them.
  * [x] need snapshot.dependent_spectra and snapshot.dependent_tags for others referencing a snapshot
  * [x] **Important:** Tags may not be deleted out of order, based on timestamp
    * [x] server side
    * [x] client side - based on DOM order? Nothing can go too far wrong due to server-side deletion validation, so this should work. Or we can use/parse the created_at 
  * [x] send `has_dependent_snapshot?` along with tags response
  * [x] display "Other data depends on this snapshot. Clone the spectrum to work from an earlier version."
  * [x] use ordering to display disabled trash icon with notice "Subsequent operations depend on this snapshot."
* [x] break out PowerTag subclass
* [x] use an icon (a thumb-tack, or a lock?) to indicate that a powertag has generated a snapshot, in the operations window
* [x] calibrations generate a new snapshot, and do not overwrite the original data.
* [x] when tag deleted, ensure next latest tag is then deletable in UI
* [x] disentangle server-side range tags from spectrum initialization
* [x] warn users with a `confirm()` that they should use 2.0 or 1.0 but not both.
* [x] better indicate loading to user
* [x] automatically test that snapshot generation refers to the correct one
* [x] and that snapshot id display in tooltip is correct
* [x] deal with the scenario that someone locks your snapshot by using it but you don't see until the page reloads. Maybe respond "locked" to the deletion command and then convert it to a locked tag, then display an error
  * [x] display spinning progress icon instead of trash, until confirmed deleted/locked
* [x] cloning shouldn't reverse tag created_at order
* [x] ensure that deleting an operation or snapshot/version checks `is_deletable?` on server side
  * [x] needs a "is_deletable()" method on client side; (needs manual and automated testing) 
  * [x] .has_snapshot should be `needs_snapshot` and should have a server-backed `has_snapshot` with a `snapshot_id`
* [x] tag.change_reference(snapshot_id)
* [x] offer to update the reference
  * [x] offer to add a reference to the "this spectrum has no snapshots"
  * [x] also account for message "spectrum has snapshots but this tag does not refer to one in particular. Choose one..."
* [x] change Tag description to return Snapshot id too
* [x] Snapshots are noted in the Operations tab at their moment in the timeline
  * [x] use a "warning" icon to indicate that a spectrum has a new snapshot -- maybe a `snapshot.is_latest?` which finds most recent snapshot with same spectrum_id ?
  * [x] Display `Spectrum 8, which this spectrum uses, has been modified. [Update] reference to latest version?` as a message in referring spectra
* [x] warn the user that a spectrum has no snapshots -- where?
* [x] cloning a spectrum clones its snapshots? It'd have to, so those snapshots wouldn't have dependents. This won't happen by default because snapshot generation happens after the `create` hook. 
  * [x] Also, the user_id of the tags and snapshots should be changed - test for this

****

**Notes:** (earlier planning)

Goal: give users easy-to-understand control over **whether modifications to a spectrum are reflected in references to that spectrum**, for example if it is included in a set, subtracted from another spectrum, or otherwise referenced. 

Note that raw data is _essentially never touched_, so what's at stake is not data loss, but how modified versions of data cause -- or don't cause -- updates in references to a spectrum.

The current proposal is to create "snapshots" or copies of a spectrum at junctures where they are used in reference elsewhere, so that the referenced data does not change unexpectedly. We should do this as lazily as possible (minimizing computation and user work), so we're proposing to auto-snapshot at the last possible moment that's still predictable and easy to understand for the user. 

Once a snapshot is made, the referring data might display something like:

    Spectrum 8, which this spectrum uses, has been modified. [Update] reference to latest version?

![img_20150903_113335](https://cloud.githubusercontent.com/assets/24359/9662931/0954d91e-5230-11e5-8b86-9a6685103791.jpg)

1. Imagine Spectrum A refers to Spectrum B, which is used in a subtraction operation.
2. So when Spectrum B is modified, does it auto-create a snapshot B1 since it is used in reference?
3. Does it assign the reference in A to the snapshot B1 automatically? Or does B get locked, and must be cloned to B1 where it can then have additional modifications added?
4. Or alternatively, does B1 get pre-emptively created at the moment A refers to it? It does seem less likely that B will be modified, which might suggest this is less efficient, but could seem to put the onus of snapshotting on the referrer, not the referree, although of course it's automated.
5. Does this system work identically whether these references are made by the author or by another user, or should we disallow users from referencing one anothers' data unless they make their own clone of it? (prefer identically, as it's more consistent)

**Another issue** -- Snapshots may work better as their own table, as _children_ of Spectrums, which just store the current frozen state, and are referenced by id from tags, like `subtract:6` for Snapshot 6. Then they won't appear in Spectrum search results, and can't be commented on, and don't have tags or un-flattened operations -- all of which simplify things. But then we have to think about how users might want to change the Snapshot a given Spectrum refers to, or make Snapshots that are of previous states. It creates some new questions if we don't just re-use the Spectrum model for these copies. 

Comments welcome!

**Comment by Stoft1** [comment]
Jeff, I'd suggest keeping data separation between users (one user can never operate on or with the original data of another user -- it must always be on/with a local copy of that data) separate from the data history mechanisms available to a user.

Eg: User-A saves a spectrum. Filters of that spectrum only exist as a list of operations to be performed before display or download but a particular set of filters can be saved as a new version (just not the modified data itself). User-B can reference the original or a filtered version, but if they want to actually use that data, they must make a copy. If they want a filtered copy, they get the original data along with the mod list for the version they wanted to use. Since they now have a copy (which has appropriate tags as to its providence for future reference) User-B's work with that data is now independent from User-A data so no residual dependencies cause backward effects on A.

So, 1) original data is protected as you said, 2) no need to save the data of filter operations (just what filter to use), 3) another user must use a local copy (with providence and filter list) and 4) original data is never affected by anything from another user.

Hope I've not missed your points.

**Comment by jywarren** [comment]
Great - actually inter-user rules are somewhat simpler to think about, and I tend to agree with you -- but do you mean that if you subtract someone else's spectrum (Spectrum A) from your own, you must make your own copy of A, say, A1, owned by yourself, before you can do that? (we might automate it, but still?)

**Comment by Stoft1** [comment]
There's actually two cases of that. 1) If B subtract's A's original, and 2) if B subtracts A's filtered data. Part of this is the question of 1) is it GUARANTEED that A's original data will ALWAYS exist and NEVER be changed/updated/replaced. If so, then B is safe to use that data -- but if B uses filtered data, then that filter might be changed by A. See the issue? To avoid all these potential dependencies, it is much much easier to just say that B is required (must) grab an make a local-to-B copy of A's data, along with the flags for filtering. Then, B is free to subtract or whatever w/o any such dependencies. This way, B is free to then delete (but not edit) a copy of A's original data if/when it is no longer useful, w/o any effect on A. Now IF you could actually guarantee that ALL of A's original data is permanent and unalterable, then the copy is not needed --- except (there's always at least one of these) if B doesn't want the original but a specific filtered version. B would then use A-original+filters and and subtract .... and then A would change a filter and B's results would change. Ah, so you say no, the A-data is unchanged but B decides on the filters -- which could work, but I suspect this sequence would be much more confusing for user B to keep track of -- especially if there were multiple filters on the A data they wanted and didn't try to keep track of everything that A did to get the spectrum display that B wanted to use. So, I'm suggesting the A-B separation and that B always makes a copy of A+filters as being 1) simpler and 2) eliminates a lot of potential dependencies. Dependencies are very difficult to program or program-around.

**Comment by jywarren** [comment]
Unfortunately there are two scenarios where there could be a change to A:

1. The extraction of numerical data from the image is itself a kind of
filter, although the original image never changes. But snapshots would fix
that, so we're good.
2. The original data could be deleted. It's hard to say if we could come up
with a system where snapshots could outlive their parent spectra. It seems
dubious. But do we forbid deletion of any data? Or any data which is
referenced anywhere?

**Comment by Stoft1** [comment]
I'd suggest that, given a substantial amount of existing original data may be eventually need to be sent to the bit bucket as just too old, a fluke, can't recall what it represents, etc, basing any scheme on assumed permanence is likely to fail. Also, any data extracted from a spectrum may also be subject to change as even calculations, corrections, calibrations etc tend to evolve and improve with time. While raw data can easily be flawed (eg. a config error) and filtering can change, at least the raw data is still the measurement source (while it still exists) and single-point-source is generally the best point to start.

Now, under the scenario of User-B making a copy and then data-A being deleted, it might not be too difficult to then tag B's copy to show that the original A-data is changed/deleted. Probably don't want the overhead of always doing this, but could, perhaps, perform the check if user B clicks on a request to check.

**Comment by Stoft1** [comment]
There is one case for 'permanent' data I can think of -- reference data which has met some standard as a reference which most may not have the ability to produce -- as opposed to the bulk of contributed data which has not met (thought it later could) some standard. The 'reference' data might be saved as 'permanent' and the general contributed data would be potentially volatile. Of course, this just again makes the case for a data handling process where it must be assumed the raw data might go away -- so create a local copy in order to use it.

**Comment by jywarren** [comment]
I'm starting to think in parallel about how to represent this to the user... frozen/snapshotted data (or whatever we ultimately decide upon) could be shown in the new "operations" table, at the point where it was snapshotted. Any later operations would appear after it. 

![screenshot 2015-09-10 at 11 10 02 pm](https://cloud.githubusercontent.com/assets/24359/9806339/72a61cde-5811-11e5-985c-058d93ca36d2.png)


This could be a vote for the mental model where to remove operations that would affect a snapshot, you have to fork off before that point -- which could also be shown in the operations table with a forking symbol: 

![screenshot 2015-09-10 at 11 13 01 pm](https://cloud.githubusercontent.com/assets/24359/9806354/8745a8c6-5811-11e5-9f68-11411c7543c6.png)

**Comment by jywarren** [comment]
OK, circling back, i'm going to start reorganizing the top of this issue and implementing it. My original thought was that we'd want to avoid making an extra copy whenever possible, but now I'm tending to think that a really predictable, consistent convention will be worth the redundancy. So, I like Dave's suggestion that any time you reference something, say with a powertag like a subtraction, we auto-generate a snapshot, owned by whomever referred to it, so they maintain control over that frozen copy from that moment in the data's history. There might be multiple snapshots from that moment -- identical, but connected to different referring spectra. Redundant, but an easy model to follow.

One other question, then -- if someone else snapshots your spectrum, then you want to delete your data, are you allowed to? By the terms of the licensing, we don't have to permit users to delete their data, since they've CC-zero licensed it. We could allow deletion UNLESS there's a snapshot relying on the data.

**Comment by jywarren** [comment]
This is complete. Whoa.
